dask_expr/merge.py

```diff
@@ -0,0 +1,226 @@
+        right = self._as_meta(self.right)
+        return left.merge(right, **self._kwargs)
+
+    def _divisions(self):
+        npartitions_left = self._as_expr(self.left).npartitions
+        npartitions_right = self._as_expr(self.right).npartitions
+        npartitions = max(npartitions_left, npartitions_right)
+        return (None,) * (npartitions + 1)
+
+    @staticmethod
+    def _as_meta(dep, nonempty=True):
+        if isinstance(dep, Expr):
+            dep = dep._meta
+        return meta_nonempty(dep) if nonempty else make_meta(dep)
+
+    @staticmethod
+    def _as_expr(dep):
+        if isinstance(dep, Expr):
+            return dep
+        return FromPandas(dep, 1)
+
+    def _simplify_down(self):
+        # Lower from an abstract expression using
+        # logic in MergeBackend.from_abstract_merge
+        backend = self.backend or MergeBackend
+        if hasattr(backend, "from_abstract_merge"):
+            return backend.from_abstract_merge(self)
+        else:
+            raise ValueError(f"{backend} not supported")
+
+
+class MergeBackend(Merge):
+    """Base merge-backend class"""
+
+    def _simplify_down(self):
+        return None
+
+    @classmethod
+    def from_abstract_merge(cls, expr: Merge) -> Expr:
+        """Return a new Exprs to perform a merge"""
+
+        # TODO:
+        #  1. Handle mixed indexed merge
+        #  2. Add multi-partition broadcast merge
+        #  3. Add/leverage partition statistics
+
+        left = expr._as_expr(expr.left)
+        right = expr._as_expr(expr.right)
+        how = expr.how
+        on = expr.on
+        left_on = expr.left_on
+        right_on = expr.right_on
+        left_index = expr.left_index
+        right_index = expr.right_index
+
+        for o in [on, left_on, right_on]:
+            if isinstance(o, Expr):
+                raise NotImplementedError()
+        if (
+            not on
+            and not left_on
+            and not right_on
+            and not left_index
+            and not right_index
+        ):
+            on = [c for c in left.columns if c in right.columns]
+            if not on:
+                left_index = right_index = True
+
+        if on and not left_on and not right_on:
+            left_on = right_on = on
+            on = None
+
+        supported_how = ("left", "right", "outer", "inner")
+        if how not in supported_how:
+            raise ValueError(
+                f"dask.dataframe.merge does not support how='{how}'."
+                f"Options are: {supported_how}."
+            )
+
+        # Check for "trivial" broadcast (single partition)
+        npartitions = max(left.npartitions, right.npartitions)
+        if (
+            npartitions == 1
+            or left.npartitions == 1
+            and how in ("right", "inner")
+            or right.npartitions == 1
+            and how in ("left", "inner")
+        ):
+            return BlockwiseMerge(left, right, **expr._kwargs)
```

Comment:
I would expect this (and general algorithm selection) to be in `Merge.simplify_down`.  Thoughts?

Reply:
I'm open to changes like this in the future.  I don't think that we're closing the door on these things by putting everything in simplify_down.  

If we're going to make a new protocol or way of doing things then I'm going to be pretty strict that it's the correct choice for the future.  If we want to reuse an existing abstraction then I'm going to be pretty relaxed.  I completely agree that in the future we may want to separate out different kinds of optimization/simplification passes.  I don't think that we have that need today, and so I'd like to keep everything in as few protocols as possible.  Once we find a need for separating things out then we'll have, I think, more information with a concrete use case to help us do this separation.

Reply:
> Users should be able to use custom logic to convert abstract expressions (like Shuffle and Merge) to other expressions. Therefore, the user needs a way to pass in this custom logic.

Also, maybe I don't have full context here.  Are there examples that you need soon like this?  Maybe for rapids work that I'm unaware of?

Reply:
>Are there examples that you need soon like this?

I'm not sure how "soon", but we absolutely want it to be easy to plug in explicit-comms shuffle/merge algorithms. We'd also like to explore custom p2p algorithms with GPU-direct storage.  I'd like to be able to just define a simple  `Merge`/`Shuffle` class that defines the necessary logic to lower from the abstract class.

Reply:
OK, good to know.  The approch here though doesn't provide that kind of registration ability though today, right?  

My sense is that the MergeBackend class today is unnecessary.  And that this logic is fundamentally the same as `Merge.simplify_down`.  If we want to add some ability to register new algorithms I'm happy with that.  I don't see that this class with this method helps us achieve that though.  

I still think that we should just do the dumb thing for now and put this in simplify_down, and then, once we have some concrete case to work with, think about how to add some kind of dispatch / registration logic.  I suspect that we'll find a simpler solution at that time.

Reply:
At the time, it may also just make sense to ask if the workers have a GPU and then import from RAPIDS directly in this code too, right?  That seems like it might be a lot simpler.

Reply:
>The approch here though doesn't provide that kind of registration ability though today, right?

Right now, one could define a `MergeBackend` class with a custom `from_abstract_merge` method, and pass it in with `backend=...`. In the future, we'd also want to implement a registration mechanism to change defaults and/or expose the algorithm in a config field.

>My sense is that the MergeBackend class today is unnecessary.

Agree that it isn't necessary yet, and even in the long run, `Merge` could just expose a `from_abstract_merge` or `lower_abstract_merge` class method.

>I still think that we should just do the dumb thing for now and put this in simplify_down, and then, once we have some concrete case to work with, think about how to add some kind of dispatch / registration logic. I suspect that we'll find a simpler solution at that time.

This is fine with me. Just being transparent that I think putting everything in `simplify_down` will probably get in the way pretty soon.